Quickstart for EKS. (#420)
bitnami/kube-prod-runtime: docs/quickstart-eks.md

# Quickstart: BKPR on Amazon Elastic Container Service for Kubernetes (Amazon EKS)

## TOC

- [Introduction](#introduction)
- [Prerequisites](#prerequisites)
- [Installation and setup](#installation-and-setup)
- [Next steps](#next-steps)
  + [Installing Kubeapps on BKPR](kubeapps-on-bkpr.md)
- [Upgrading](#upgrading)
- [Teardown and cleanup](#teardown-and-cleanup)
- [Further reading](#further-reading)

## Introduction

This document walks you through setting up an Amazon Elastic Container Service for Kubernetes (Amazon EKS) cluster and installing the Bitnami Kubernetes Production Runtime (BKPR) on it.

## Prerequisites

* [Amazon AWS account](https://aws.amazon.com/)
* [Amazon CLI](https://aws.amazon.com/cli/)
* [`eksctl`](https://aws.amazon.com/blogs/opensource/eksctl-eks-cluster-one-command/)
* [Kubernetes CLI](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
* [BKPR installer](install.md)
* [kubecfg](https://github.com/ksonnet/kubecfg/releases)
* [jq](https://stedolan.github.io/jq/)

### DNS requirements

> In addition to the requirements listed above, a domain name is also required for setting up Ingress endpoints to services running in the cluster. The specified domain name can be a top-level domain (TLD) or a subdomain. In either case you have to manually [set up the NS records](#step-3-configure-domain-registration-records) for the specified TLD or subdomain so as to delegate DNS resolution queries to an Amazon Route 53 hosted zone created and managed by BKPR.  This is required in order to generate valid TLS certificates.

Reviewer: s/(#step-3-configure-domain-registration-records)/(#step-4-configure-domain-registration-records)/

Author: Done

## Installation and setup

### Step 1: Set up the cluster

> In this section, you will deploy an Amazon Elastic Container Service for Kubernetes (Amazon EKS) cluster using the `eksctl` tool`.

Reviewer 2: ```suggestion
In this section, you will deploy an Amazon Elastic Container Service for Kubernetes (Amazon EKS) cluster using the `eksctl` tool.
```

Author: Done.

> * Make sure you have a working Amazon CLI environment. Please refer to [Configuring the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html) if in doubt.

Reviewer 2: ```suggestion
* Make sure you have a working Amazon CLI environment. Refer to [Configuring the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html) for more information.
```

Author: Done.

* Configure the following environment variables:

  ```bash
  export BKPR_DNS_ZONE=my-domain.com
  export AWS_EKS_USER=my-user@example.com
  export AWS_EKS_CLUSTER=my-eks-cluster
  export AWS_EKS_K8S_VERSION=1.11
  ```

  - `BKPR_DNS_ZONE` specifies the DNS suffix for the externally-visible websites and services deployed in the cluster.
  - `AWS_EKS_USER` specifies the email address used in requests to Let's Encrypt.
  - `AWS_EKS_CLUSTER` specifies the name of the AKS cluster.

>   - `AWS_EKS_K8S_VERSION` specifies the version of Kubernetes to use for creating the cluster. The [BKPR Kubernetes version support matrix](../README.md#kubernetes-version-support-matrix-for-bkpr-10) lists the base Kubernetes versions supported by BKPR.

Reviewer: s/(../README.md#kubernetes-version-support-matrix-for-bkpr-10)/(../README.md#kubernetes-version-support-matrix-for-bkpr)

Author: Done.

* Create the EKS cluster:

  ```bash
  eksctl create cluster --name=${AWS_EKS_CLUSTER} \
                        --color=fabulous \
                        --nodes=3 \
                        --version=${AWS_EKS_K8S_VERSION}
  ```

  Provisioning an EKS cluster can take a long time to complete. Please be patient while the request is being processed.

* Configure `kubectl` to use the new cluster:

  ```bash
  aws eks update-kubeconfig --name=${AWS_EKS_CLUSTER}
  ```

* Verify that your cluster is up and running:

  ```bash
  kubectl get nodes
  ```

### Step 2: Configure a User Pool in AWS Cognito

In order to authenticate users for applications running atop EKS, BKPR requires a User Pool in AWS Cognito to be configured.

> If you already have a working User Pool in AWS Cognito that you would like to use for authenticating users, you will need to retrieve its ID in the form `${AWS_REGION}-${USER_POOL_ID}`, and write it down. Using this `Pool Id`, skip at the end of this section.

Reviewer: Shall we request at this point to export it as AWS_COGNITO_USER_POOL_ID (as it says below). Also, what is the end of this section? move to the next one? something else?

Author: Done

Else, If you are new to using BKPR on EKS, or if you want to create a new User Pool in AWS Cognito, please follow these steps:

1. Browse to the [Cognito](https://console.aws.amazon.com/cognito/) module in the AWS Console

> 2. Click on the `Manage User Pools` button

Reviewer 2: I suggest using **bold** text when referring to UI elements. **Manager User Pools** in this case. IMO code formatting should generally be used when referring to a `command`, `VARIABLE`, etc.

Side-note: You don't explicitly need to increment numbered lists in markdown. You can write them plainly as

```
1. Step 1
1. Next Step
1. Another Step
```

which will be rendered correctly as shown below

1. Step 1
1. Next Step
1. Another Step

Author: I can't use the 1. trick because a paragraph break resets the numbering.

3. Click on the `Create a user pool` button on the top-right corner

> 4. Enter a valid `Pool name` like `eks-test`

Reviewer 2: Consider merging these steps into one: f.e.

Navigate to **Manage User Pools > Create a user pool** and specify a **Pool name**, fe. _eks-test_

5. Click on the `Step through settings` button
6. In the `Attributes` section leave all the checkboxes unticked as shown below:

> <p align="center"><img src="eks/1-Attributes.png" width=840/></p>

Reviewer: Please, use markdown for the images as well (```![Alt Text](url)```)

Author: The images look too big when using markdown, hence the reason for pure HTML tags (controlling size).

> 7. Click on the `Next step` button

Reviewer 2: I think this step should be merged with the previous step as in we can write the previous step as follows:

In the **Attributes** section leave all the check-boxes unticked and go to the **Next step**:

The same suggestion goes for other instances as well

8. In the `Policies` section leave all the checkboxes unticked and select `Only allow administrators to create users` as shown below:

<p align="center"><img src="eks/2-Policies.png" width=840/></p>

9. Click on the `Next step` button
10. In the `MFA and verifications` section select `No verification` as shown below:

<p align="center"><img src="eks/3-MFA.png" width=840/></p>

11. Click on the `Next step` button
12. In the `Message customizations` section do not change anything and click on the `Next step` button
13. In the `Tags` section you can add additional tags for your EKS cluster, if needed
14. Click on the `Next step` button
15. In the `Devices` section do not change anything and click on the `Next step` button
16. In the `App clients` section do not change anything and click on the `Next step` button
17. In the `Triggers` section do not change anything and click on the `Next step` button
18. In the `Review` section you will see a quick summary of the user pool that will be created, like the onw shown below:

<p align="center"><img src="eks/4-Review.png" width=840/></p>

19. Click on the `Create pool` button to create the user pool

> 20. The user pool will be created and the `General settings` page will be displayed, where you can find the User Pool ID and ARN:

Reviewer 2: I suggest merging a lot of these steps

Author: Done.

<p align="center"><img src="eks/5-User-Pool.png" width=840/></p>

Take note of the `Pool Id` and export its value:

  ```bash
  export AWS_COGNITO_USER_POOL_ID=eu-central-1_sHSdWT6VL
  ```

In order to access protected resources which require authentication like Prometheus, Kibana or Grafana, you will need to configure users in the newly created user pool. The next steps highlight how to create a test user which can be used to access these protected resources:

1. Browse to the [Cognito](https://console.aws.amazon.com/cognito/) module in the AWS Console
2. Click on the `Manage User Pools` button
3. Click on the button for the user pool you just created
4. Click on the `Users and groups` option from the menu on the left

> 5. Click on the `Create user` button

Reviewer 2: I would suggest changing this entire sequence to:

2. Navigate to **Manage User Pools > _YOUR_USER_POOL_ > Users and Groups > Create user**

Congratulations! You can now deploy your applications on the Kubernetes cluster and BKPR will help you manage and monitor them effortlessly.

## Next steps

- [Installing Kubeapps on BKPR](kubeapps-on-bkpr.md)

## Upgrading

### Step 1: Update the installer

Follow the [installation guide](install.md) to update the BKPR installer binary to the latest release.

### Step 2: Edit `kubeprod-manifest.jsonnet`

Edit the `kubeprod-manifest.jsonnet` file that was generated by `kubeprod install` and update the version referred in the `import` statement. For example, the following snippet illustrates the changes required in the `kubeprod-manifest.jsonnet` file if you're upgrading to version `v1.3.0` from version `v1.2.0`.

```diff
 // Cluster-specific configuration
-(import "https://releases.kubeprod.io/files/v1.2.0/manifests/platforms/eks.jsonnet") {
+(import "https://releases.kubeprod.io/files/v1.3.0/manifests/platforms/eks.jsonnet") {
  config:: import "kubeprod-autogen.json",
  // Place your overrides here
 }
```

### Step 3: Perform the upgrade

> Re-run the `kubeprod install` command, from the [Deploy BKPR](#step-2-deploy-bkpr) step, in the directory containing the existing `kubeprod-autogen.json` and updated `kubeprod-manifest.jsonnet` files.

Reviewer: s/#step-2-deploy-bkpr/#step-3-deploy-bkpr

Author: Done.
